code/config/lib/Loot/Config/Env.hs

```diff
@@ -0,0 +1,361 @@
+        -- maybe there are cases when it is mostly not :thinking:
+        -- My idea is to add "isSecure" flag to 'EnvValue' typelass to resolve
+        -- this problem.
+        "Failed to parse environmental variable \
+        \" +| errKey |+ "=" +| maybe "-" build errValue |+ "\
+        \: " +| errMessage |+ ""
+
+-- | Pretty-print a 'EnvParseError'.
+parseErrorPretty :: EnvParseError -> String
+parseErrorPretty = pretty
+
+instance Exception EnvParseError where
+    displayException = fmt . build
+
+-- | Environmental value parser.
+--
+-- Use 'fail' to report parsing errors, and 'noValue' to indicate
+-- that value is left uninitialized.
+newtype Parser a = Parser
+    { runParser :: MaybeT (Except Text) a
+      -- ^ Parsing result, Maybe layer designates result presence, while Except
+      -- layer contains parsing errors.
+    } deriving (Functor, Applicative, Monad)
+
+instance MonadFail Parser where
+    fail = Parser . lift . throwError . fromString
+
+-- | Leave value uninitialized in config.
+noValue :: Parser a
+noValue = Parser mzero
+
+-- | Describes a way to parse an item appearing in config.
+class EnvValue a where
+    -- | Parse a variable value.
+    parseEnvValue :: Maybe String -> Parser a
+
+-- | Apply the given parser to an environmental value if it present,
+-- otherwise leave the configuration option uninitialized.
+--
+-- This is what most parsers usually do.
+withPresent :: (String -> Parser a) -> Maybe String -> Parser a
+withPresent parser = \case
+    Nothing -> noValue
+    Just val -> parser val
+
+-- | Value parser based on 'IsString' instance.
+parseStringEnvValue :: IsString a => Maybe String -> Parser a
+parseStringEnvValue = withPresent $ pure . fromString
+
+-- | Value parser based on 'Read' instance.
+autoParseEnvValue :: Read a => Maybe String -> Parser a
+autoParseEnvValue = withPresent $ \arg -> case reads arg of
+    -- This is similar to what e.g. optparse-applicative does,
+    -- the errors from 'eitherRead' are not too beautiful
+    [(r, "")] -> return r
+    [(_, s)]  -> fail $ "Extra trailing input '" <> s <> "'"
+    []        -> fail "Failed to read value"
+    _         -> fail "Value is ambigous"
+
+-- | Value parser for numbers with overflow checks.
+parseBoundedNumEnvValue
+    :: forall a. (Bounded a, Integral a)
+    => Maybe String -> Parser a
+parseBoundedNumEnvValue val = do
+    int <- autoParseEnvValue @Integer val
+    if | int < fromIntegral (minBound @a) -> fail "Numeric underflow"
+       | int > fromIntegral (maxBound @a) -> fail "Numeric overflow"
+       | otherwise -> pure (fromIntegral int)
+
+-- | Value parser based on 'Aeson.FromJSON' instance.
+aesonParseEnvValue :: Aeson.FromJSON a => Maybe String -> Parser a
+aesonParseEnvValue =
+    withPresent $ either fail pure . Aeson.eitherDecode . encodeUtf8
+
+-- | Options which define the expected format of environmental variables.
+data ParseOptions = ParseOptions
```

Comment:
Let’s provide a helper for easily making `keyBuilder`s that prefix the name. Realistically, in the absolute majority of cases you want all your env variables prefixed with `<APP_NAME>_`.